_posts/2023-07-24-single-vs-multi-tenant.md

```diff
@@ -0,0 +1,190 @@
+|-----------------------------------|
+|                                   |
+|    +---------------------------+  |
+|    |      Application Logic    |  |
+|    +---------------------------+  |
+|                |                 |
+|                v                 |
+|    +---------------------------+  |
+|    |        Database           |  |
+|    |  (Shared Among Tenants)   |  |
+|    +---------------------------+  |
+|                                   |
+-------------------------------------
+
+In this diagram, the multi-tenant application's logic is shared among all tenants, meaning that there is a single instance of the application logic serving all customers. However, each tenant's data is stored separately in a shared database, ensuring that data remains isolated and secure for each customer.
+
+The shared database is designed in such a way that each tenant's data is logically partitioned, allowing them to access only their specific data without any access to data belonging to other tenants. This approach enables efficient resource utilization and cost savings, as multiple tenants can share the same infrastructure.
+
+
+## When to choose Multi-Tenant Model
+
+Multi-tenant architectures are best suited for scenarios where resource efficiency, cost-effectiveness, and scalability are top priorities. Here are some situations where multi-tenant might be the preferred choice:
+
+- Software as a Service (SaaS): SaaS providers commonly use multi-tenant architectures to offer software applications to a large number of customers efficiently. It allows them to serve numerous clients with a single software instance.
+
+- Cost Savings: Small and medium-sized businesses, startups, or organizations with limited budgets can benefit from multi-tenant architectures as they can share the cost of resources, infrastructure, and maintenance among multiple tenants.
+
+- Scalability: Applications that experience varying levels of demand can leverage the scalability of multi-tenant setups. 
+Resources can be dynamically allocated and adjusted based on the changing needs of different tenants.
+
+- Collaborative Environments: Applications that encourage collaboration and interaction between users, like project management tools or social networking platforms, can benefit from the shared nature of multi-tenant architectures.
+
+- Rapid Deployment: Multi-tenant setups allow for quick deployment of the application to multiple customers, reducing the time and effort required for individual setups.
+
+Here's an example of an application that would thrive with a multi tenant architecture:
+
+ An email service provider offers email accounts to thousands of users. By using a multi-tenant architecture, the provider can efficiently manage resources and offer cost-effective email solutions to a large user base. Each user's data remains separate and secure, while the underlying infrastructure is shared among multiple users.
+
+## Multi-Tenant Hybrid
```

Comment:
This is the really interesting bit for me.  Most systems end up being hybrid. The question is what point in the data flow do you change from shared to isolated?

Reply:
For SAAS products, it seems like most products are mainly on hybrid. The goal of this blogpost is for people to aware there's multiple types you can choose from. Like your comment above , single instance for each client seems like an expensive option, but it is a valid option tho if they see there is a need for their use case.